go-controller/pkg/util/gateway-init.go

```diff
@@ -39,68 +39,80 @@ func getLocalSystemID() (string, error) {
-		"nb_global", ".", "external-ids:gateway-lock=\"\"", "--", "set",
-		"nb_global", ".", "external_ids:gateway-lock="+localSystemID)
-	if err != nil {
-		return fmt.Errorf("Failed to set gateway-lock "+
-			"stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+	// Convert \r\n to \n to support Windows line endings
+	stdout = strings.Replace(strings.TrimSpace(stdout), "\r\n", "\n", -1)
+	gatewayRouters := strings.Split(stdout, "\n")
+	if len(gatewayRouters) == 0 {
+		return "", nil, fmt.Errorf("failed to get default gateway router (no routers found)")
 	}
-	return nil
+	parts := strings.Fields(gatewayRouters[0])
+	var ip net.IP
+	for _, p := range parts {
+		const forceTag string = "lb_force_snat_ip="
+		if strings.HasPrefix(p, forceTag) {
+			ip = net.ParseIP(p[len(forceTag):])
+			break
+		}
+	}
+	if ip == nil {
+		return "", nil, fmt.Errorf("failed to parse gateway router %q IP", parts[0])
+	}
+	return parts[0], ip, nil
 }
 
-func unlockNBForGateways() {
-	stdout, stderr, err := RunOVNNbctl("--", "set", "nb_global", ".",
-		"external-ids:gateway-lock=\"\"")
-	if err != nil {
-		logrus.Errorf("Failed to delete lock for gateways, "+
-			"stdout: %q, stderr: %q, error: %v", stdout, stderr, err)
+func ensureGatewayPortAddress(portName string) (string, *net.IPNet, error) {
+	routerMac, routerIP, _ := GetPortAddresses(portName, false)
+	if routerMac == "" || routerIP == "" {
+		// Create the gateway switch port in 'join' if it doesn't exist yet
+		stdout, stderr, err := RunOVNNbctl("--wait=sb",
+			"--may-exist", "lsp-add", "join", portName,
+			"--", "--if-exists", "clear", "logical_switch_port", portName, "dynamic_addresses",
+			"--", "lsp-set-addresses", portName, "dynamic")
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to add logical switch "+
+				"port %s, stdout: %q, stderr: %q, error: %v",
+				portName, stdout, stderr, err)
+		}
+		// Wait for its addresses
+		routerMac, routerIP, err = GetPortAddresses(portName, false)
+		if err != nil {
+			return "", nil, fmt.Errorf("Error while waiting for addresses "+
+				"for gateway switch port %q: %v", portName, err)
+		}
 	}
-}
 
-func generateGatewayIP() (string, error) {
-	// All the routers connected to "join" switch are in 100.64.1.0/24
-	// network and they have their external_ids:connect_to_join set.
-	stdout, stderr, err := RunOVNNbctl("--data=bare", "--no-heading",
-		"--columns=network", "find", "logical_router_port",
-		"external_ids:connect_to_join=yes")
+	// Grab the 'join' switch prefix to add to our gateway router's IP
+	joinSubnet, stderr, err := RunOVNNbctl("--if-exists", "get",
+		"logical_switch", "join", "other-config:subnet")
 	if err != nil {
-		logrus.Errorf("Failed to get logical router ports which connect to "+
-			"\"join\" switch, stdout: %q, stderr: %q, error: %v",
-			stdout, stderr, err)
-		return "", err
+		return "", nil, fmt.Errorf("Failed to get 'join' switch subnet: "+
+			"%s, stderr: %q, %v", joinSubnet, stderr, err)
 	}
```

Comment:
Perhaps you can skip printing `joinsubnet` at L99 above since if we have failed it might be empty anyways.